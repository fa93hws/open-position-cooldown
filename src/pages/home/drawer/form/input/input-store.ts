import { makeObservable, observable, action } from 'mobx';

export class InputStore {
  value = '';

  hasError = false;

  shouldValidate = false;

  constructor(
    private readonly validators: readonly ((val: string) => boolean)[],
  ) {
    makeObservable(this, {
      value: observable.ref,
      setValue: action,
      hasError: observable.ref,
    });
  }

  setValue(val: string) {
    this.value = val;
    if (!this.shouldValidate) {
      return;
    }
    this.hasError = this.validators.some((validator) => !validator(this.value));
  }

  startValidate() {
    this.shouldValidate = true;
    this.setValue(this.value);
  }
}
