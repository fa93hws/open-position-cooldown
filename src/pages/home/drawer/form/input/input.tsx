import * as React from 'react';
import { TextField } from '@material-ui/core';
import { observer } from 'mobx-react';

import { InputStore } from './input-store';

export type ExposedInputProps = {
  label: string;
};

type InputProps = ExposedInputProps & {
  value: string;
  error: boolean;
  onChange(val: string): void;
  onBlur(): void;
};

export const Input = React.memo((props: InputProps) => {
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    props.onChange(event.target.value);
  };
  return (
    <TextField
      label={props.label}
      onChange={handleChange}
      onBlur={props.onBlur}
      value={props.value}
      error={props.error}
      fullWidth
      color="primary"
      type="text"
      variant="outlined"
      size="small"
    />
  );
});

export function createInput(validators: readonly ((val: string) => boolean)[]) {
  const store = new InputStore(validators);
  const onChange = (val: string) => store.setValue(val);
  const onBlur = () => store.startValidate();

  const Component = observer((props: ExposedInputProps) => (
    <Input
      value={store.value}
      error={store.hasError}
      onChange={onChange}
      onBlur={onBlur}
      label={props.label}
    />
  ));
  return [Component, store] as const;
}
