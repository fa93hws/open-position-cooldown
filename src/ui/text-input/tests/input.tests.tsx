import * as React from 'react';
import {
  render,
  cleanup,
  fireEvent,
  getByPlaceholderText,
} from '@testing-library/react';

import { createInput, Input } from '../input';

describe('Input', () => {
  beforeEach(cleanup);

  it('renders with value', () => {
    expect(
      <Input
        value="hello world"
        error={false}
        label="label"
        onChange={jest.fn()}
        onBlur={jest.fn()}
      />,
    ).toMatchRenderedSnapshot();
  });

  it('renders with value with error', () => {
    expect(
      <Input
        value="hello world"
        error={true}
        label="label"
        onChange={jest.fn()}
        onBlur={jest.fn()}
      />,
    ).toMatchRenderedSnapshot();
  });

  it('is assign the value on change', () => {
    const [Component, store] = createInput([]);
    const { container } = render(<Component placeholder="foo" />);
    fireEvent.change(getByPlaceholderText(container, 'foo'), {
      target: { value: '123' },
    });
    expect(store.value).toEqual('123');
  });

  it('starts validating when blurred', () => {
    const [Component, store] = createInput([jest.fn().mockReturnValue(false)]);
    const { container } = render(<Component placeholder="foo" />);
    expect(store.hasError).toEqual(false);
    fireEvent.blur(getByPlaceholderText(container, 'foo'));
    expect(store.hasError).toEqual(true);
  });
});
