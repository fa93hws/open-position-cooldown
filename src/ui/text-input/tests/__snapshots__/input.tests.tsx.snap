// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders with value 1`] = `
<div
  class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
>
  <label
    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiInputLabel-outlined MuiFormLabel-filled"
    data-shrink="true"
  >
    label
  </label>
  <div
    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
  >
    <input
      aria-invalid="false"
      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
      type="text"
      value="hello world"
    />
    <fieldset
      aria-hidden="true"
      class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
    >
      <legend
        class="PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4"
      >
        <span>
          label
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`Input renders with value with error 1`] = `
<div
  class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
>
  <label
    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiInputLabel-outlined Mui-error Mui-error MuiFormLabel-filled"
    data-shrink="true"
  >
    label
  </label>
  <div
    class="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
  >
    <input
      aria-invalid="true"
      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
      type="text"
      value="hello world"
    />
    <fieldset
      aria-hidden="true"
      class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
    >
      <legend
        class="PrivateNotchedOutline-legendLabelled-7 PrivateNotchedOutline-legendNotched-8"
      >
        <span>
          label
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
